### ISSUE TITLE
Clippy suggests broken code when modifying union fields behind ManuallyDrop

### ISSUE THREAD

**Issue by abgros** [rust-lang/rust-clippy#14386]
### Summary

Modifying a union field that is behind ManuallyDrop incorrectly triggers the "immediately dereferencing a reference" lint, causing Clippy to suggest broken code.

### Reproducer

```rust
use std::mem::ManuallyDrop;

struct Data {
	num: u64,
}

union DataWithPadding {
	data: ManuallyDrop<Data>,
	with_padding: [u8; size_of::<Data>()],
}

fn main() {
	let mut a = DataWithPadding {
		with_padding: [0; size_of::<DataWithPadding>()],
	};
	unsafe {
		(*(&raw mut a.data)).num = 42;
		// Clippy suggests `(a.data).num = 42;`, which doesn't work

		println!("{:?}", a.with_padding);
	}
}
```

### Version

```text
rustc 1.87.0-nightly (8c392966a 2025-03-01)
binary: rustc
commit-hash: 8c392966a013fd8a09e6b78b3c8d6e442bc278e1
commit-date: 2025-03-01
host: x86_64-pc-windows-msvc
release: 1.87.0-nightly
LLVM version: 20.1.0
```

**Comment by abgros** [comment]
Turns out that adding a single dereference fixes things:
```rust
(*a.data).num = 42;
```

**Comment by samueltardieu** [comment]
This is a bug nevertheless. Thanks for the report.

@rustbot claim
